File: src/Shapes.java
import java.util.concurrent.ThreadLocalRandom;

class Shapes {
    private int[][][] current;
    private int figureNumber = 0;
    private int[][][][] figures =
            {
                    {//I
                            {
                                    {0, 0}, {0, 1}, {0, 2}, {0, 3}
                            },
                            {
                                    {-1, 1}, {0, 1}, {1, 1}, {2, 1}
                            }
                    },
                    {//L
                            {
                                    {1, 0}, {0, 0}, {0, 1}, {0, 2}
                            },
                            {
                                    {-1, 0}, {-1, 1}, {0, 1}, {1, 1}
                            },
                            {
                                    {1, 0}, {1, 1}, {1, 2}, {0, 2}
                            },
                            {
                                    {-1, 1}, {0, 1}, {1, 1}, {1, 2}
                            }
                    },
                    {//J
                            {
                                    {0, 0}, {0, 1}, {0, 2}, {1, 2}
                            },
                            {
                                    {-1, 1}, {0, 1}, {1, 1}, {1, 0}
                            },
                            {
                                    {0, 0}, {1, 0}, {1, 1}, {1, 2}
                            },
                            {
                                    {-1, 2}, {-1, 1}, {0, 1}, {1, 1}
                            }
                    },
                    {//O
                            {
                                    {0, 0}, {0, 1}, {1, 1}, {1, 0}
                            }
                    },
                    {//S
                            {
                                    {1, 0}, {1, 1}, {0, 1}, {0, 2}
                            },
                            {
                                    {-1, 1}, {0, 1}, {0, 2}, {1, 2}
                            },
                    },
                    {//T
                            {
                                    {0, 0}, {0, 1}, {0, 2}, {1, 1}
                            },
                            {
                                    {0, 0}, {0, 1}, {1, 1}, {-1, 1}
                            },
                            {
                                    {0, 0}, {0, 1}, {0, 2}, {-1, 1}
                            },
                            {
                                    {-1, 1}, {0, 1}, {0, 2}, {1, 1}
                            }
                    },
                    {//Z
                            {
                                    {0, 0}, {0, 1}, {1, 1}, {1, 2}
                            },
                            {
                                    {-1, 2}, {0, 2}, {0, 1}, {1, 1}
                            }
                    }
            };
    private int state = 0;
    private int next = ThreadLocalRandom.current().nextInt(0,7);
    private int nextShapeNumber;

    void rotate() {
        state = (state + 1) % current.length;
    }

    int[][] getRotateCoordinates(int offsetX, int offsetY) {
        int[][] result = new int[current[0].length][2];
        for (int i = 0; i < result.length; i++) {
            result[i][0] = current[(state + 1) % current.length][i][0] + offsetY;
            result[i][1] = current[(state + 1) % current.length][i][1] + offsetX;
        }
        return result;
    }

    int[][] getPosition(int offsetX, int offsetY) {
        int[][] result = new int[current[0].length][2];
        for (int i = 0; i < result.length; i++) {
            result[i][0] = current[state][i][0] + offsetY;
            result[i][1] = current[state][i][1] + offsetX;
        }
        return result;
    }

    int getShapeNumber() {
        return figureNumber;
    }

    void setRandom() {
        state = 0;
        int temp = next;
        next = ThreadLocalRandom.current().nextInt(0, 700000000) / 100000000;
        nextShapeNumber = next + 1;
        figureNumber = temp + 1;
        current = figures[temp];
    }
    int getNextShapeNumber() {
        return nextShapeNumber;
    }
    int[][] getNextShape() {
        return figures[next][0];
    }
}

File: src/game/Shapes.kt
package game

import java.util.concurrent.ThreadLocalRandom

internal class Shapes {
    private var current = arrayOf(arrayOf(intArrayOf()))
    var shapeNumber = 0
        private set
    private val figures = arrayOf(
            arrayOf(//I
                    arrayOf(intArrayOf(0, 0), intArrayOf(0, 1), intArrayOf(0, 2), intArrayOf(0, 3)),
                    arrayOf(intArrayOf(-1, 0), intArrayOf(0, 0), intArrayOf(1, 0), intArrayOf(2, 0))),
            arrayOf(//L
                    arrayOf(intArrayOf(1, 0), intArrayOf(0, 0), intArrayOf(0, 1), intArrayOf(0, 2)),
                    arrayOf(intArrayOf(-1, 0), intArrayOf(-1, 1), intArrayOf(0, 1), intArrayOf(1, 1)),
                    arrayOf(intArrayOf(1, 0), intArrayOf(1, 1), intArrayOf(1, 2), intArrayOf(0, 2)),
                    arrayOf(intArrayOf(-1, 0), intArrayOf(0, 0), intArrayOf(1, 0), intArrayOf(1, 1))),
            arrayOf(//J
                    arrayOf(intArrayOf(0, 0), intArrayOf(0, 1), intArrayOf(0, 2), intArrayOf(1, 2)),
                    arrayOf(intArrayOf(-1, 1), intArrayOf(0, 1), intArrayOf(1, 1), intArrayOf(1, 0)),
                    arrayOf(intArrayOf(0, 0), intArrayOf(1, 0), intArrayOf(1, 1), intArrayOf(1, 2)),
                    arrayOf(intArrayOf(-1, 1), intArrayOf(-1, 0), intArrayOf(0, 0), intArrayOf(1, 0))),
            arrayOf(//O
                    arrayOf(intArrayOf(0, 0), intArrayOf(0, 1), intArrayOf(1, 1), intArrayOf(1, 0))),
            arrayOf(//S
                    arrayOf(intArrayOf(1, 0), intArrayOf(1, 1), intArrayOf(0, 1), intArrayOf(0, 2)),
                    arrayOf(intArrayOf(-1, 0), intArrayOf(0, 0), intArrayOf(0, 1), intArrayOf(1, 1))),
            arrayOf(//T
                    arrayOf(intArrayOf(0, 0), intArrayOf(0, 1), intArrayOf(0, 2), intArrayOf(1, 1)),
                    arrayOf(intArrayOf(0, 0), intArrayOf(0, 1), intArrayOf(1, 1), intArrayOf(-1, 1)),
                    arrayOf(intArrayOf(0, 0), intArrayOf(0, 1), intArrayOf(0, 2), intArrayOf(-1, 1)),
                    arrayOf(intArrayOf(-1, 0), intArrayOf(0, 0), intArrayOf(0, 1), intArrayOf(1, 0))),
            arrayOf(//Z
                    arrayOf(intArrayOf(0, 0), intArrayOf(0, 1), intArrayOf(1, 1), intArrayOf(1, 2)),
                    arrayOf(intArrayOf(-1, 1), intArrayOf(0, 1), intArrayOf(0, 0), intArrayOf(1, 0))
            )
    )
    private var state = 0
    private var next = ThreadLocalRandom.current().nextInt(0, 7)
    var nextShapeNumber: Int = 0
        private set
    val nextShape: Array<IntArray>
        get() = figures[next][0]
    val statesCount: Int
        get() = current.size

    fun rotate() {
        state = (state + 1) % current.size
    }

    fun getWidth(st: Int): Int =
            when (shapeNumber - 1) {
                0 -> if (st == 0) 4 else 1
                3 -> 2
                else -> if (st % 2 == 0) 3 else 2
            }

    fun getRotateCoordinates(offsetX: Int, offsetY: Int): Array<IntArray> {
        val result = Array(current[0].size) { IntArray(2) }
        for (i in result.indices) {
            result[i][0] = current[(state + 1) % current.size][i][0] + offsetY
            result[i][1] = current[(state + 1) % current.size][i][1] + offsetX
        }
        return result
    }

    fun getPosition(offsetX: Int, offsetY: Int, st: Int = this.state): Array<IntArray> {
        val result = Array(current[0].size) { IntArray(2) }
        for (i in result.indices) {
            result[i][0] = current[st][i][0] + offsetY
            result[i][1] = current[st][i][1] + offsetX
        }
        return result
    }

    fun setRandom() {
        state = 0
        val temp = next
        next = ThreadLocalRandom.current().nextInt(0, 700000000) / 100000000
        nextShapeNumber = next + 1
        shapeNumber = temp + 1
        current = figures[temp]
    }

    fun removeSame(element: IntArray, array: MutableList<IntArray>) {
        val it = array.iterator()
        while (it.hasNext()) {
            if (element.contentEquals(it.next())) it.remove()
        }
    }

    fun getBottomCells(st: Int): Array<IntArray> {
        val result = mutableListOf<IntArray>()
        result += getPosition(0, 1, st)
        for (element in current[st]) {
            removeSame(element, result)
        }
        return result.toTypedArray()
    }


    fun getNeighboursCells(st: Int): Array<IntArray> {
        val result = mutableListOf<IntArray>()
        result += getPosition(0, 1, st)
        result += getPosition(-1, 0, st)
        result += getPosition(1, 0, st)
        for (element in current[st]) {
            removeSame(element, result)
        }
        return result.toTypedArray()
    }
}
